engine/common/rpc/convert/events.go

```diff
@@ -0,0 +1,177 @@
+		if err != nil {
+			return nil, fmt.Errorf("could not convert event payload from CCF to Json: %w", err)
+		}
+		event.Payload = convertedPayload
+		return &event, nil
+	case execproto.EventEncodingVersion_JSON_CDC_V0:
+		return &event, nil
+	default:
+		return nil, fmt.Errorf("invalid encoding format %d", inputVersion)
+	}
+}
+
+// MessagesToEventsFromVersion converts a slice of protobuf messages to a slice of flow.Events, converting
+// the payload encoding from CCF to JSON if the input version is CCF
+func MessagesToEventsFromVersion(l []*entities.Event, version execproto.EventEncodingVersion) ([]flow.Event, error) {
+	events := make([]flow.Event, len(l))
+	for i, m := range l {
+		event, err := MessageToEventFromVersion(m, version)
+		if err != nil {
+			return nil, fmt.Errorf("could not convert event at index %d from format %d: %w",
+				m.EventIndex, version, err)
+		}
+		events[i] = *event
+	}
+	return events, nil
+}
+
+// ServiceEventToMessage converts a flow.ServiceEvent to a protobuf message
+func ServiceEventToMessage(event flow.ServiceEvent) (*entities.ServiceEvent, error) {
+	bytes, err := json.Marshal(event.Event)
+	if err != nil {
+		return nil, fmt.Errorf("cannot marshal service event: %w", err)
+	}
+
+	return &entities.ServiceEvent{
+		Type:    event.Type.String(),
+		Payload: bytes,
+	}, nil
+}
+
+// MessageToServiceEvent converts a protobuf message to a flow.ServiceEvent
+func MessageToServiceEvent(m *entities.ServiceEvent) (*flow.ServiceEvent, error) {
+	rawEvent := m.Payload
+	eventType := flow.ServiceEventType(m.Type)
+	se, err := flow.ServiceEventJSONMarshaller.UnmarshalWithType(rawEvent, eventType)
+
+	return &se, err
+}
+
+// ServiceEventListToMessages converts a slice of flow.ServiceEvents to a slice of protobuf messages
+func ServiceEventListToMessages(list flow.ServiceEventList) (
+	[]*entities.ServiceEvent,
+	error,
+) {
+	entities := make([]*entities.ServiceEvent, len(list))
+	for i, serviceEvent := range list {
+		m, err := ServiceEventToMessage(serviceEvent)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert service event at index %d to message: %w", i, err)
+		}
+		entities[i] = m
+	}
+	return entities, nil
+}
+
+// ServiceEventsToMessages converts a slice of flow.ServiceEvents to a slice of protobuf messages
+func MessagesToServiceEventList(m []*entities.ServiceEvent) (
+	flow.ServiceEventList,
+	error,
+) {
+	parsedServiceEvents := make(flow.ServiceEventList, len(m))
+	for i, serviceEvent := range m {
+		parsedServiceEvent, err := MessageToServiceEvent(serviceEvent)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse service event at index %d from message: %w", i, err)
+		}
+		parsedServiceEvents[i] = *parsedServiceEvent
+	}
+	return parsedServiceEvents, nil
+}
+
+// CcfPayloadToJsonPayload converts a CCF-encoded payload to a JSON-encoded payload
+func CcfPayloadToJsonPayload(p []byte) ([]byte, error) {
+	if len(p) == 0 {
+		return p, nil
+	}
```

Comment:
this is the only change to the converter logic. This was needed otherwise ccf will return an eof error.

Reply:
I wonder if this change is needed. 

Is a valid CcfPayload always not empty? In other words, an empty payloads CcfPayload must be invalid? In that case, then the default behavior of returning an error (EOF error) is expected, the caller is better to receive an error instead of an empty bytes. Right?

cc @fxamacker

Reply:
> Is a valid CcfPayload always not empty? 

Yes, valid CCF message is never empty.

> the caller is better to receive an error instead of an empty bytes. Right?

Yes, I agree we should return error on empty data.  JSON-CDC decoder currently returns error on empty data as well.